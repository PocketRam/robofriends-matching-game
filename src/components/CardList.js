import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Card from './Card';

class CardList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      matchingCards: [],
      flippedCard: [],
      openedCard: [],
      numberOfMoves: 0
    };

    this.onCardClick = this.onCardClick.bind(this);
  }

  onCardClick(event) {
    if (this.state.flippedCard.length && this.state.openedCard.length) {
      return;
    }

    const flippedCard = [event.target.id, event.target.alt];
    this.setState({ flippedCard: flippedCard });

    if (this.state.openedCard.length) {
      this.doCardsMatch(event.target.id, event.target.alt);
    } else {
      this.setState({ flippedCard: [] });
      this.setState({ openedCard: flippedCard });
    }
  }

  doCardsMatch(flippedCardId, flippedCardAlt) {
    const self = this;

    setTimeout(function() {
      let matchingCards = self.state.matchingCards.slice(0);
      const cardsMatch = self.state.openedCard[0] !== flippedCardId && 
        self.state.openedCard[1] === flippedCardAlt;
      const newFlippedCard = [flippedCardId, flippedCardAlt];
      
      if (cardsMatch) {
        matchingCards.push(self.state.openedCard);
        matchingCards.push(newFlippedCard);
        self.setState({ matchingCards: matchingCards });
      }

      self.setState({ flippedCard: [] });
      self.setState({ openedCard: [] });
      self.setState({ numberOfMoves: self.state.numberOfMoves + 1 });
    }, 1000);
  }

  render() {
    const { robots } = this.props;
    const { flippedCard, matchingCards, openedCard, numberOfMoves } = this.state;
    
    return (
      <ul className="card-list">
        <li>Number of moves: {numberOfMoves}</li>
       {
          robots.map((robot, i) => {
            return <Card key={i} 
                         id={`robot${i}-${robot}`} 
                         alt={`robot${robot}`}
                         flippedCard={flippedCard}
                         matchingCards={matchingCards}
                         openedCard={openedCard}
                         onCardClick={this.onCardClick} />;
          })
        }
      </ul>
    );
  }
}

CardList.propTypes = {
  robots: PropTypes.array
};

export default CardList;