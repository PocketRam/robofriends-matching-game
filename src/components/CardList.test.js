import React from 'react';
import { shallow } from 'enzyme';
import CardList from './CardList';

const matchRobots = [1];

it('renders without crashing', () => {
  expect(shallow(<CardList robots={matchRobots}/>)).toMatchSnapshot();
});