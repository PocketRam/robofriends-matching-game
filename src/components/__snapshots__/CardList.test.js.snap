// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardList
    robots={
        Array [
            1,
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li>
          Number of moves: 
          0
</li>,
        Array [
          <Card
            alt="robot1"
            flippedCard={Array []}
            id="robot0-1"
            matchingCards={Array []}
            onCardClick={[Function]}
            openedCard={Array []}
/>,
        ],
      ],
      "className": "card-list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Number of moves: ",
            0,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Number of moves: ",
          0,
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "alt": "robot1",
          "flippedCard": Array [],
          "id": "robot0-1",
          "matchingCards": Array [],
          "onCardClick": [Function],
          "openedCard": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li>
            Number of moves: 
            0
</li>,
          Array [
            <Card
              alt="robot1"
              flippedCard={Array []}
              id="robot0-1"
              matchingCards={Array []}
              onCardClick={[Function]}
              openedCard={Array []}
/>,
          ],
        ],
        "className": "card-list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Number of moves: ",
              0,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Number of moves: ",
            0,
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "alt": "robot1",
            "flippedCard": Array [],
            "id": "robot0-1",
            "matchingCards": Array [],
            "onCardClick": [Function],
            "openedCard": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
